src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs

```diff
@@ -366,8 +377,174 @@ public override ExtendedSpecialType ExtendedSpecialType
+                        !methodSymbol.ReturnsVoid ||
+                        methodSymbol.ParameterCount != 1) // PROTOTYPE: Should we accept more than one parameter (in case new versions add more info)?
+                    {
+                        return null; // PROTOTYPE: Test this code path
+                    }
+
+                    return new ReceiverParameterSymbol(this, methodSymbol.Parameters[0]);
+                }
+
+                MethodSymbol getMarkerMethodSymbol()
+                {
+                    Debug.Assert(!uncommon.MarkerMethod.IsNil);
+
+                    foreach (var member in GetMembers(WellKnownMemberNames.ExtensionMarkerMethodName))
+                    {
+                        if (member is PEMethodSymbol candidate && candidate.Handle == uncommon.MarkerMethod)
+                        {
+                            return candidate;
+                        }
+                    }
+
+                    throw ExceptionUtilities.Unreachable();
+                }
+            }
+        }
+
+        private sealed class ReceiverParameterSymbol : RewrittenParameterSymbol
+        {
+            private readonly PENamedTypeSymbol _containingType;
+
+            public ReceiverParameterSymbol(PENamedTypeSymbol containingType, ParameterSymbol originalParameter) :
+                base(originalParameter)
+            {
+                _containingType = containingType;
+            }
+
+            public override Symbol ContainingSymbol
+            {
+                get { return _containingType; }
+            }
+        }
+
+        public override MethodSymbol TryGetCorrespondingExtensionImplementationMethod(MethodSymbol method)
+        {
+            Debug.Assert(this.IsExtension);
+            Debug.Assert(method.IsDefinition);
+            Debug.Assert(method.ContainingType == (object)this);
+
+            if (this.ContainingType is null)
+            {
+                return null; // PROTOTYPE: Test this code path
+            }
+
+            if (!method.IsStatic && ExtensionParameter is null)
+            {
+                return null; // PROTOTYPE: Test this code path
+            }
+
+            var uncommon = GetUncommonProperties().lazyExtensionInfo;
+
+            if (uncommon.LazyImplementationMap is null)
+            {
+                Interlocked.CompareExchange(ref uncommon.LazyImplementationMap, new ConcurrentDictionary<MethodSymbol, MethodSymbol>(Roslyn.Utilities.ReferenceEqualityComparer.Instance), null);
+            }
+
+            return uncommon.LazyImplementationMap.GetOrAdd(
+                method,
+                static (method, @this) =>
+                {
+                    foreach (var member in @this.ContainingType.GetMembers(SourceExtensionImplementationMethodSymbol.GetImplementationName(method)))
+                    {
+                        if (member is not MethodSymbol { HasSpecialName: true, IsStatic: true } candidate)
```

Comment:
Should we check the accessibility of the implementation method (should match the skeleton method's)?